### PR TITLE
Mark keys for Config required only with InstalledSite

### PR DESCRIPTION
**Pull request by luceos**
Fixes an issue with the $_SERVER object having no value in cli, this is used in the cli install command.

### CODE REVIEW

**Comment by askvortsov1** [comment]
Unit tests need to be adjusted to explicitly require the url key in the test that checks for it.

File: src/Foundation/Config.php [hunk 1]
```diff
@@ -42,7 +40,7 @@ public function inMaintenanceMode(): bool
         return $this->data['offline'] ?? false;
     }
 
-    private function requireKeys(...$keys)
+    public function requireKeys(...$keys)
```
Comment (askvortsov1) [inline src/Foundation/Config.php:43]: Not sure how I feel about mutating the singleton... I would prefer that we pass in required keys to the constructor, or make this method return a clone of the config object

File: src/Foundation/Site.php [hunk 2]
```diff
@@ -25,7 +25,10 @@ public static function fromPaths(array $paths)
         date_default_timezone_set('UTC');
 
         if (! static::hasConfigFile($paths->base)) {
-            return new UninstalledSite($paths, $_SERVER['REQUEST_URI']);
+            return new UninstalledSite(
+                $paths,
+                Arr::get($_SERVER, 'REQUEST_URI')
```
Comment (askvortsov1) [inline src/Foundation/Site.php:30]: Is this change necessary for this PR?

**Review by askvortsov1 — CHANGES_REQUESTED** [review]

File: src/Foundation/Site.php (hunk 2 above)
Comment (luceos) [reply src/Foundation/Site.php:30]: Without this change the CLI still fails, because the original error is that $_SERVER does not contain that key:

```
PHP Notice:  Undefined index: REQUEST_URI in /home/forge/nightly.flarum.site/nightly/vendor/flarum/core/src/Foundation/Site.php on line 28
```

This is **the** fix. But because in the code the URL cannot be NULL we needed deeper changes.

File: src/Foundation/Config.php (hunk 1 above)
Comment (luceos) [reply src/Foundation/Config.php:43]: Agreed, will update and fix tests in one swoop.

Comment (luceos) [reply src/Foundation/Config.php:43]: This won't work. The construct of this class is fired from the site.php which is too easily accessible from the outside. Either we drop outside mutation of the required keys from outside (demanding them based on a flag perhaps $installed as argument) or we use the above change.

File: src/Foundation/Site.php (hunk 2 above)
Comment (askvortsov1) [reply src/Foundation/Site.php:30]: Makes sense. Let's keep that then!

File: src/Foundation/Config.php (hunk 1 above)
Comment (askvortsov1) [reply src/Foundation/Config.php:43]: In this case, let's temporarily remove requiring _any_ keys, and revisit validation later? The previous config object didn't have external validation, so adding it now isn't necessary (plus, the choice of 'url' as the main attribute to enforce needs more discussion in retrospect)

Comment (luceos) [reply src/Foundation/Config.php]: Patched tests, removed method.

File: src/Foundation/Site.php (hunk 2 above)
Comment (franzliedke) [inline src/Foundation/Site.php:30]: What do we need the value for at runtime for the installer? Since this is for the `UninstalledSite`, we could just provide a bogus value when `$_SERVER['REQUEST_URI']` isn't set?

I'd find that cleaner in this case.

Comment (franzliedke) [reply src/Foundation/Site.php:30]: When reviewing the old code, it looks like the old version must have had `url` set to `null` in the CLI, so a bogus value should work fine and be less invasive?

Comment (askvortsov1) [reply src/Foundation/Site.php:30]: Hmm, generally I'm not the biggest fan of bogus values, but in this case, we could just put in the nightly site's literal url I suppose? And since we'd expect a url to be provided on CLI install, I suppose we wouldn't run into this issue for actual usage?

Comment (franzliedke) [reply src/Foundation/Site.php:30]: I would literally put a bogus value there - something that signals that this is intentionally bogus, because irrelevant. `doesnotmatter.tld` or `bogus.url` maybe?

Comment (franzliedke) [reply src/Foundation/Site.php:30]: A comment can help explain the choice, of course.

Comment (askvortsov1) [reply src/Foundation/Site.php:30]: Wait, in the core source code, or in the nightly site config?

Comment (franzliedke) [reply src/Foundation/Site.php:30]: Core source code. When we pass a base URL to the `UninstalledSite` in `Site.php`.

Comment (askvortsov1) [reply src/Foundation/Site.php:30]: Ah alright I get what you're getting at.

Small followup that I should have thought of earlier: is there a reason why we require `url`, and only `url`?

Comment (franzliedke) [reply src/Foundation/Site.php:30]: Good question. Haven't thought that true, I guess. :wink:

Comment (franzliedke) [reply src/Foundation/Site.php:30]: Or actually... it's the one that's used far more than any of the other ones, but obviously the other ones must be used somewhere, too.

Comment (askvortsov1) [reply src/Foundation/Site.php:30]: Tbh, that's mostly why I preferred being able to pass in required keys per-site, or maybe even a flag. How do you feel about the mutation-based approach that @luceos  had in the original PR?

Comment (franzliedke) [reply src/Foundation/Site.php:30]: Meh. At that point, we could easily do two classes again.

File: src/Foundation/Site.php [hunk 3]
```diff
@@ -25,7 +25,10 @@ public static function fromPaths(array $paths)
         date_default_timezone_set('UTC');
 
         if (! static::hasConfigFile($paths->base)) {
-            return new UninstalledSite($paths, $_SERVER['REQUEST_URI']);
+            return new UninstalledSite(
+                $paths,
+                Arr::get($_SERVER, 'REQUEST_URI', 'http://localhost')
```
Comment (askvortsov1) [inline src/Foundation/Site.php:30]: I like `bogus.url` a bit more as it's clearer that this doesn't actually mean anything. But either way, we'd want a comment explaining that this is only needed for validation at this stage, and doesn't actually do anything

**Review by askvortsov1 — CHANGES_REQUESTED** [review]

**Review by askvortsov1 — APPROVED** [review]

**Review by tankerkiller125 — APPROVED** [review]
